php/EE/RevertableStepProcessor.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+namespace EE;
+
+/**
+ * RevertibleStepProcessor
+ *
+ * This class is used to ensure that a series of steps have been executed successfully.
+ * If any one step fails while executing, All executed steps will be reverted.
+ */
+class RevertableStepProcessor {
+
+	/** @var array Contains array of steps */
+	private $steps = [];
+
+	/** @var int Keeps track of steps executed. All items in $steps till this index have been executed */
+	private $execution_index = 0;
+
+	/**
+	 * Adds a new step.
+	 *
+	 * @param callable $up_step Callable that will be called when step is to be executed
+	 * @param callable $down_step Callable that will be called when step is to be reverted
+	 * @param string $context Context of step. It will be used to display error.
+	 */
+	public function add_step( callable $up_step, callable $down_step, string $context = null ) {
+		$this->steps[] = [
+			'up' => $up_step,
+			'down' => $down_step,
+			'context' => $context,
+		];
+
+		return $this; // Returns this to enable method chaining
+	}
+
+	/**
+	 * Adds new step and executes pending steps.
+	 *
+	 * @param callable $up_step Callable that will be called when step is to be executed
+	 * @param callable $down_step Callable that will be called when step is to be reverted
+	 * @param string $context Context of step. It will be used to display error.
+	 */
+	public function execute_step( callable $up_step, callable $down_step, string $context = null ) {
+		$this->add_step( $up_step, $down_step, $context );
+		return $this->execute();
+	}
+
+	/**
+	 * Executes all pending steps. Reverts the steps if any one step throws error.
+	 */
+	public function execute() {
+		for ( $i = $this->execution_index ; $i < count( $this->steps ); $i++ ) {
```

Comment:
> I think here you can just use foreach with $this->steps and avoid extra variable $i usage plus count function

According the logic written, we don't have to execute all steps. We only have to execute the steps that have not been executed. `execution_index` keeps track of which steps have been executed. Hence we need to execute steps from index `$this->execution_index` till end of `$this->steps`

> Please try to avoid repetitive operation like here you're using count() function in the loop on each iteration it will calculate $this->steps

Yes. Will fix this.

Reply:
>We only have to execute the steps that have not been executed.

Well for this you're just skipping all the steps in loop to exectue from specific index which can be done in foreach loop also. 

But I think here the better approach would be to first get the array you want from the index you need by using `array_slice` and then you can use that array.